File: lib/tests/test_rd_grid_coarse.cpp
/**
 * Tests coarse cell groups.
 *
 * A coarse cell group identifies a contiguous block of cells that the
 * simulator treats as a single lumped cell. This is encoded by the
 * CORSNUM keyword in an EGRID file: one int per cell, where 0 means "not in
 * any coarse group" and positive values are group identifiers.
 */

#include <catch2/catch.hpp>
#include <resdata/rd_grid.hpp>
#include <resdata/rd_kw.hpp>
#include <resdata/rd_kw_magic.hpp>
#include <vector>

#include "grid_fixtures.hpp"
#include "tmpdir.hpp"

using namespace Catch;
using namespace Matchers;

TEST_CASE_METHOD(Tmpdir, "Load EGRID with coarse cell groups", "[unittest]") {
    GIVEN("An EGRID file on disc with two coarse cell groups") {
        const int nx = 3, ny = 3, nz = 3;
        const int size = nx * ny * nz;

        // Assign two coarse groups. 0 means the cell is not in a
        // coarse group. Group 1 covers the two cells along i at (0,0,0)
        // and (1,0,0); group 2 covers two cells in the last k layer.
        std::vector<int> corsnum(size, 0);
        corsnum[0] = 1;
        corsnum[1] = 1;
        corsnum[size - 1] = 2;
        corsnum[size - 2] = 2;

        auto filename = dirname / "CORSNUM.EGRID";
        auto grid =
            load_egrid_with_coarse_groups(filename, nx, ny, nz, corsnum.data());

        REQUIRE(grid != nullptr);

        THEN("The grid reports that coarse cells are present") {
            REQUIRE(rd_grid_have_coarse_cells(grid.get()));
            REQUIRE(rd_grid_get_num_coarse_groups(grid.get()) == 2);
        }

        THEN("Cells in a coarse group are flagged as such, others are not") {
            REQUIRE(rd_grid_cell_in_coarse_group1(grid.get(), 0));
            REQUIRE(rd_grid_cell_in_coarse_group1(grid.get(), 1));
            REQUIRE_FALSE(rd_grid_cell_in_coarse_group1(grid.get(), 2));
            REQUIRE(rd_grid_cell_in_coarse_group1(grid.get(), size - 1));
            REQUIRE(rd_grid_cell_in_coarse_group1(grid.get(), size - 2));
        }

        THEN("The coarse group objects are accessible by index") {
            REQUIRE(rd_grid_iget_coarse_group(grid.get(), 0) != nullptr);
            REQUIRE(rd_grid_iget_coarse_group(grid.get(), 1) != nullptr);
        }

        THEN("The grid reports no fracture cells") {
            REQUIRE(rd_grid_get_active_fracture_index1(grid.get(), 0) == -1);
            REQUIRE(rd_grid_get_active_fracture_index1(grid.get(), size / 2) ==
                    -1);
            REQUIRE(rd_grid_get_global_index1F(grid.get(), 0) == -1);
        }

        THEN("The grid reports a consistent (empty) LGR tree") {
            REQUIRE(rd_grid_test_lgr_consistency(grid.get()));
        }

        THEN("rd_grid_init_actnum_data returns the actnum") {
            std::vector<int> actnum(size);
            rd_grid_init_actnum_data(grid.get(), actnum.data());
            for (int i = 0; i < size; i++)
                REQUIRE(actnum[i] ==
                        (rd_grid_cell_active1(grid.get(), i) ? 1 : 0));
        }
    }
}

TEST_CASE_METHOD(Tmpdir, "Load EGRID with non-consecutive coarse group numbers",
                 "[unittest]") {
    // CORSNUM values are allowed to have gap, we return null for skipped slots.
    GIVEN("An EGRID file whose CORSNUM uses groups 1 and 3 but not 2") {
        const int nx = 3, ny = 3, nz = 3;
        const int size = nx * ny * nz;

        std::vector<int> corsnum(size, 0);
        corsnum[0] = 1;
        corsnum[1] = 1;
        corsnum[size - 1] = 3;
        corsnum[size - 2] = 3;

        auto filename = dirname / "CORSNUM_SPARSE.EGRID";
        auto grid =
            load_egrid_with_coarse_groups(filename, nx, ny, nz, corsnum.data());
        REQUIRE(grid != nullptr);

        THEN("rd_grid_get_num_coarse_groups reports the highest group number "
             "even though an intermediate group is unused") {
            REQUIRE(rd_grid_have_coarse_cells(grid.get()));
            REQUIRE(rd_grid_get_num_coarse_groups(grid.get()) == 3);
        }

        THEN("rd_grid_iget_coarse_group returns a non-null object for used "
             "group numbers and null for the skipped one") {
            REQUIRE(rd_grid_iget_coarse_group(grid.get(), 0) != nullptr);
            REQUIRE(rd_grid_iget_coarse_group(grid.get(), 1) == nullptr);
            REQUIRE(rd_grid_iget_coarse_group(grid.get(), 2) != nullptr);
        }
    }
}

File: tests/rd_tests/test_rd_grid_coarse.py
from __future__ import annotations

import pytest

from ._grid_fixtures import load_egrid_with_coarse_groups


@pytest.fixture
def two_groups_grid(tmp_path):
    nx, ny, nz = 3, 3, 3
    size = nx * ny * nz
    corsnum = [0] * size
    corsnum[0] = 1
    corsnum[1] = 1
    corsnum[size - 1] = 2
    corsnum[size - 2] = 2
    return load_egrid_with_coarse_groups(
        tmp_path / "CORSNUM.EGRID",
        nx,
        ny,
        nz,
        corsnum,
    )


def test_that_grid_with_two_coarse_groups_reports_coarse_cells_present(two_groups_grid):
    assert two_groups_grid.coarse_groups() == 2


def test_that_cells_in_a_coarse_group_are_flagged_as_such(two_groups_grid):
    size = two_groups_grid.get_global_size()
    assert two_groups_grid.in_coarse_group(global_index=0)
    assert two_groups_grid.in_coarse_group(global_index=1)
    assert not two_groups_grid.in_coarse_group(global_index=2)
    assert two_groups_grid.in_coarse_group(global_index=size - 1)
    assert two_groups_grid.in_coarse_group(global_index=size - 2)


def test_that_grid_with_two_coarse_groups_reports_no_fracture_cells(two_groups_grid):
    size = two_groups_grid.get_global_size()
    assert two_groups_grid.get_active_fracture_index(global_index=0) == -1
    assert two_groups_grid.get_active_fracture_index(global_index=size // 2) == -1
    assert two_groups_grid.get_global_index1F(0) is None


def test_that_actnum_export_matches_per_cell_active_status(two_groups_grid):
    actnum = list(two_groups_grid.export_actnum())
    for i in range(two_groups_grid.get_global_size()):
        expected = 1 if two_groups_grid.active(global_index=i) else 0
        assert actnum[i] == expected


def test_that_non_consecutive_coarse_group_numbers_report_highest_group(tmp_path):
    # CORSNUM uses groups 1 and 3 but not 2. The grid reports the
    # highest group number even though group 2 is unused.
    nx, ny, nz = 3, 3, 3
    size = nx * ny * nz
    corsnum = [0] * size
    corsnum[0] = 1
    corsnum[1] = 1
    corsnum[size - 1] = 3
    corsnum[size - 2] = 3

    grid = load_egrid_with_coarse_groups(
        tmp_path / "CORSNUM_SPARSE.EGRID",
        nx,
        ny,
        nz,
        corsnum,
    )
    assert grid.coarse_groups() == 3
